### PR TITLE
Allow user to set timezone

### PR DESCRIPTION
**Pull request by Timshel**
Hey,

This adds the ability for a user to set a Timezone which will replace the default one from the browser.

Worked on it mainly because Firefox with resist fingerprinting (Cf https://github.com/element-hq/element-web/issues/25006) set the timezone to UTC, and I'm not aware of an easy way to whitelist only specific domains.

Edit: probably should have searched more before coding but there is supposed to be a whitelist in Firefox RFP but for now I have issues with it :sweat_smile:. And this might be useful in other cases.

## Checklist

-   [ ] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Comment by t3chguy** [comment]
Sorry wrong button

**Comment by Timshel** [comment]
Not sure of the process but the PR is in a relative good state and working well.

Set it as WIP since:
 - I'm aware of a test failing since the properties screen now has more element.
 - Might need more translation ?

But wanted some feedback before spending more time on it :).

**Comment by t3chguy** [comment]
I've requested review from @matrix-org/product but they are quite busy so it may take some time, it is their decision what features land

**Comment by daniellekirkwood** [comment]
I'd like @americanrefugee 's input on the design of the button and it's layout. In terms of functionality I would accept this enhancement if/when implemented correctly 

Here's how I see it in the Netlify build. Let me know if I'm actually looking at the wrong thing.
Screenshot 2024-07-16 at 12 35 19

**Review by americanrefugee — CHANGES_REQUESTED** [review]
The functionality of the Netlify screenshot if correct, but the design should look like this:
https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=97-15874&t=whVB5NajX00nqnN5-4

**Comment by Timshel** [comment]
Made a short recording of the field in action:

Recording 2024-07-16 at 15 19 17

**Comment by Timshel** [comment]
@americanrefugee I don't have access even with an account to the Figma file.

**Comment by americanrefugee** [comment]
Sorry @Timshel I didn't realize at first that you're not an Element employee - which is why I can't give you access to our design files.

However, here are screenshots of how the feature should look like, as well as dropdown menu states:
Dropdown
Settings_ General - PC

As for the functionality, I agree that it's useful to be able to type to filter dropdown menu results. However, the user shouldn't actually be typing anything into the field.

**Comment by t3chguy** [comment]
Such a Dropdown component is being worked on in https://github.com/element-hq/compound-web/pull/204 and should be reusable here once it lands

**Comment by Timshel** [comment]
Ok I'll switch to having the dropdown justified on the left with the label on top.

> I agree that it's useful to be able to type to filter dropdown menu results. However, the user shouldn't actually be typing anything into the field.

I'll mention first that it's reusing the same dropdown that the one used at the top of the page for `Application language`. 

But I'm not sure what you mean by not "typing anything into the field" do you mean not visible ?
The filtering is essentially a search, without the "search" term visible the user won't be able to see if he made a typo, and it will make deleting the search term quite tricky.

If you mean it to work like the classic list with keyboard navigation then I would mention that there is over 400 values so I believe filtering is quite important here.

**Comment by americanrefugee** [comment]
I suppose if the user starts typing then it can follow the regular form field component:

Type=Text, State=Typing, Dark=Off

And here is the Clear / X button:

Button_ Clear


- By default, the selected menu item label should appear
- If typing, the content are replaced with a cursor and "Clear" (X) button
   - The list should update as the user types
- If the user clicks the x / clear button, the original label re-appears and all menu items are shown

Something like that?

**Comment by Timshel** [comment]
@americanrefugee adding the Clear/ X would make sense, but I believe there is some conflict with the down arrow which indicate that it's a dropdown (probably why it's not already present). 

As I mentioned I'm just reusing an existing component, so any change would impact multiple existing fields.
To be honest I believe this is a separate issue and should not be addressed here; and since this handling is already used in the same page I don't believe it should block this PR.

**Comment by americanrefugee** [comment]
Ok, reuse the existing component. Don't wanna block anything :)

**Comment by Timshel** [comment]
@americanrefugee do you have a recommendation for the dropdown width ?

By default, it matches the width of the selected option (The container has the class `mx_SettingsSubsection_contentStretch`).
For the other dropdown (`Application dropdown` ) it matches the hint width (`The app will reload after selecting another language`).

Added the current Timezone in the `Browser default` placeholder/option using the Timezone short format defined as:

> Short localized form (e.g.: PST, GMT-8)

Cf https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat

**Comment by americanrefugee** [comment]
> do you have a recommendation for the dropdown width ?

360 pixels

**Comment by Timshel** [comment]
@americanrefugee page look look like [snapshots/settings/preferences-user-settings-tab.spec.ts](https://github.com/matrix-org/matrix-react-sdk/blob/a04d0858ec5aebe9abdf8c3d15d7c8a35ebd67b8/playwright/snapshots/settings/preferences-user-settings-tab.spec.ts/Preferences-user-settings-tab-should-be-rendered-properly-1-linux.png)
Set the `360px` using `min-width` to prevent wrapping if the application language description is wider.

**Comment by Timshel** [comment]
Hey,
I suppose someone need to review the code ? :)

**Comment by t3chguy** [comment]
@daniellekirkwood given @americanrefugee is unavailable how can we progress on this?

**Comment by Timshel** [comment]
Any news ?

**Comment by americanrefugee** [comment]
> Any news ?

LGTM. Is there anything else do you need from me atm?

**Comment by florianduros** [comment]
> > Any news ?
> 
> LGTM. Is there anything else do you need from me atm?

Can you approve the PR ? Thanks :)

**Comment by florianduros** [comment]
@Timshel Can you merge develop in the PR in order to solve the conflict and make it up to date ? I'll review it after.

**Review by americanrefugee — APPROVED** [review]

**Comment by Timshel** [comment]
I rebased but the Playwright tests should fail since I'm unable to run those (and update the snapshot); it fails with :

```
Test timeout of 30000ms exceeded while setting up "user".
```

I tested on `develop` too and had the same error.

**Comment by florianduros** [comment]
> I rebased but the Playwright tests should fail since I'm unable to run those (and update the snapshot); it fails with :
> 
> ```
> Test timeout of 30000ms exceeded while setting up "user".
> ```
> 
> I tested on `develop` too and had the same error.

Why are you unable to run the playwright tests?

**Comment by Timshel** [comment]
It's failing in: 

```
       at ../element-web-test.ts:230

      228 |     user: async ({ pageWithCredentials: page, credentials }, use) => {
      229 |         await page.goto("/");
    > 230 |         await page.waitForSelector(".mx_MatrixChat", { timeout: 30000 });

```

Raised the timeout to `60000` but same error.

**Comment by florianduros** [comment]
I think you have this error because the local instance of synapse used by the tests is not here. Normally, it's launched automatically when you launch our playwright tests.

Can you reinstall playwright and also put all the output of `yarn run test:playwright playwright/e2e/settings/preferences-user-settings-tab.spec.ts`

**Comment by Timshel** [comment]
The synapse instance appears to be running, the output from the run : https://gist.github.com/Timshel/a4ee0cdac04f7ded5fb4098cb4f74597
I had a quick look and homeserver-stdout.log / homeserver-stderr.log does not appear to contain any errors.

I wiped the `node_modules` and use `yarn cache clean` before installing again but no change.

**Comment by florianduros** [comment]
Do you have an ED/EW running when you launch the playwright tests? 
- If no => start it
- If yes => stop it

(seems weird but trust me :p)

**Comment by Timshel** [comment]
Not sure what you mean by:
 - ed: editor ?
  - ew: element web ?
  
Kept only a shell open to run the tests and no change.

Unless you want to debug the issue, might be easier, if you don't mind, if you commit the updated snapshot ? (you should have commit right to the feature branch of the PR). 

Sorry to ask it but for now I'm not working on any other modification so I would prefer not to spend the time to dig the issue (I'm using Playwright in another project which control some docker container without any issue).

**Comment by florianduros** [comment]
Ho sorry, EW => Element Web.
I'll update the snapshots then after the review

File: src/i18n/strings/fr.json
Comment (florianduros) [inline src/i18n/strings/fr.json:2636]: Please don't edit the i18n files except `en_EN.json`, the translation are handled by element and the community  https://github.com/element-hq/element-web/blob/develop/docs/translating.md

File: src/settings/Settings.tsx
Comment (florianduros) [inline src/settings/Settings.tsx:654]: In `en_EN.json`, the keys is `settings|preferences|user_timezone`.
To be avoid this kind of mistake, you can use `yarn i18n` (https://github.com/matrix-org/matrix-react-sdk/blob/develop/package.json#L36). It will populate the missing translation in `en_EN.json`.

File: src/TimezoneHandler.ts
Comment (florianduros) [inline src/TimezoneHandler.ts:22]: Can you add documentation to the file methods? Thanks

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:292]: Can you use instead our [dropdown component from our design system](https://compound.element.io/?path=/docs/compound-web_dropdown--docs)

**Comment by Timshel** [comment]
I had missed the errors from `Typescript Syntax Check` fixing those now.

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:242]: Can you add docs here to explain why we need to do it?

**Review by florianduros — CHANGES_REQUESTED** [review]
First quick review.

There are some typescript errors, missing docs and we can use a dropdown component from our design system.
Also, the PR needs to reach the 80% of code coverage in order to be merged. You are near, some extra tests are needed :)

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (Timshel) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:292]: Unless I'm mistaken the dropdown in question does not have a search and with over 400 elements I think the search is quite useful.

**Comment by florianduros** [comment]
Please request a review when the changes are made (top right button in the *Reviewers* section)

**Comment by Timshel** [comment]
Ok so the issue with the Playwright tests was just that I had not updated the  `element-web` project.

Added some test but since I don't believe Sonar run by default no idea of the updated %.

**Comment by florianduros** [comment]
Great for playwright!

Sonar is running against the jest tests, the playwright tests are not included in the code coverage.
The PR has reached 75.9% (https://github.com/matrix-org/matrix-react-sdk/pull/12775/checks?check_run_id=29364727755)
You are near :)

File: src/TimezoneHandler.ts
Comment (florianduros) [inline src/TimezoneHandler.ts]: Missing documentation for theses methods too.

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx]: I would to like to not see this array operations in the render of the component. To not be called at every render.

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:308]: This casting should not be necessary. `timezones` needs to have a correct here instead to be casted.

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [inline src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx]: The result of `TimezoneHandler.shortBrowserTimezone()` should be a variable of the default timezone translation.
The sentence can be structured differently in other languages.

```suggestion
        const browserTimezoneLabel = `_t("settings|preferences|default_timezone", { timezone:  TimezoneHandler.shortBrowserTimezone()})`
```

https://github.com/element-hq/element-web/blob/develop/docs/translating-dev.md#adding-variables-inside-a-string

**Review by florianduros — CHANGES_REQUESTED** [review]

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (Timshel) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:308]: Do you have a way to do this without a cast ?
All examples I can find in the code base are using a cast (Ex: https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/views/location/LiveDurationDropdown.tsx#L77)

File: src/TimezoneHandler.ts
Comment (Timshel) [reply src/TimezoneHandler.ts]: Not sure what to write can you give the description you want ?

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (Timshel) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx]: I don't know what I'm missing but with : 

```
"default_timezone": "Browser default (%(timezone))",
````

And 

```
_t("settings|preferences|default_timezone", { timezone: TimezoneHandler.shortBrowserTimezone()})
```

I have no interpolation and end up with `Browser default (%(timezone))`

Comment (florianduros) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx]: You missed the `s`. 

> Add the variable inside the string. The syntax for variables is %(variable)s. Please note the s at the end. The name of the variable has to match the previous used name.

`"default_timezone": "Browser default (%(timezone)s)",`

File: src/TimezoneHandler.ts
Comment (florianduros) [reply src/TimezoneHandler.ts]:
```suggestion
/**
* Set in the settings the given timezone
* @timezone
*/
export function setUserTimezone(timezone: string): Promise<void> {
    return SettingsStore.setValue(USER_TIMEZONE_KEY, null, SettingLevel.DEVICE, timezone);
}

/**
* Return all the available timezones
*/
export function getAllTimezones(): string[] {
    return Intl.supportedValuesOf("timeZone");
}

/**
* Return the current timezone in a short human readable way 
*/
export function shortBrowserTimezone(): string {
    return new Intl.DateTimeFormat(undefined, { timeZoneName: "short" })
        .formatToParts(new Date())
        .find((x) => x.type === "timeZoneName")?.value ?? "GMT";
}
```

File: src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx
Comment (florianduros) [reply src/components/views/settings/tabs/user/PreferencesUserSettingsTab.tsx:308]: Okay, so let's keep it

**Comment by Timshel** [comment]
Managed to update the playwright snapshot even if the other test is failing with a timeout (the test was not reported as failing on the ci ...).

**Comment by Timshel** [comment]
Hey sorry, with some jest test failing locally (on `develop` too) I did not look at why the playwright test started failing :(.
It appears I had unlinked `matrix-react-sdk` from my local `element-web` :(.

Hopefully this should be ok now, there is still the `Typescript Syntax Check` failing but no idea why, it's running without issue locally and the reported [errors](https://github.com/matrix-org/matrix-react-sdk/actions/runs/10618814755/job/29463793045) are clearly not limited to the change in this PR.

**Comment by florianduros** [comment]
A js-sdk PR broke our CI, so don't worry about the current state of the ts errors.
I will take a look at the test, test it manually today or next week

**Comment by Timshel** [comment]
@florianduros Just seen your latest push, I was thinking of adding to the playwright test (seen those default in other tests):

```
test.use({
  locale: 'en-GB',
  timezoneId: 'Europe/London',
});
````

This can be set globally too in the `playwright.config.ts`, but I believe it would be worth its own PR.

**Comment by florianduros** [comment]
Perfect it works!
Thanks for the contribution and not giving up! Well done!

**Review by florianduros — APPROVED** [review]
Perfect it works!
Thanks for the contribution and not giving up! Well done!

**Pull request merged**
